File: ios/Runner/OpenedFileService.swift
import Flutter
import UIKit

class OpenedFileService: NSObject {
    static let shared = OpenedFileService()

    override private init() { }

    private var channel: FlutterMethodChannel!
    private var initiallyOpenedFile: String?

    func register(with binaryMessenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(name: "opened_file", binaryMessenger: binaryMessenger)

        channel.setMethodCallHandler({
            [weak self](call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            switch call.method {
                case "getInitiallyOpenedFile":
                    let file = self?.initiallyOpenedFile
                    self?.initiallyOpenedFile = nil

                    result(file)
                    break
                default:
                    result(FlutterMethodNotImplemented)
                    break
            }
        })
    }

    func initiallyOpenedFile(url: URL) {
        initiallyOpenedFile = try? String(contentsOf: url)
    }

    func openedFile(url: URL) {
        _ = url.startAccessingSecurityScopedResource()

        channel.invokeMethod("onOpenedFile", arguments: try? String(contentsOf: url))

        url.stopAccessingSecurityScopedResource()
    }
}

File: android/app/src/main/kotlin/jozkar/mladez/OpenedFileService.kt
package jozkar.mladez

import java.io.BufferedReader
import java.io.InputStreamReader

import android.content.Intent
import android.content.Context

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

object OpenedFileService {
    private lateinit var channel: MethodChannel
    private var initiallyOpenedFile: String? = null

    fun register(binaryMessenger: BinaryMessenger) {
        channel = MethodChannel(binaryMessenger, "opened_file")
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "getInitiallyOpenedFile" -> {
                    val file = initiallyOpenedFile
                    initiallyOpenedFile = null

                    result.success(file)
                }
                else -> result.notImplemented()
            }
        }
    }

    fun initiallyOpenedFile(context: Context, intent: Intent) {
        readFile(context, intent)?.let {
            initiallyOpenedFile = it
        }
    }

    fun openedFile(context: Context, intent: Intent) {
        readFile(context, intent)?.let {
            channel.invokeMethod("onOpenedFile", it)
        }
    }

    private fun readFile(context: Context, intent: Intent): String? {
        val uri = intent.data

        if (uri != null) {
            val reader = BufferedReader(InputStreamReader(context.getContentResolver().openInputStream(uri)))
            val stringBuilder = StringBuilder()
            var line: String? = reader.readLine()

            while (line != null) {
                stringBuilder.append(line)
                line = reader.readLine()
            }

            reader.close()

            return stringBuilder.toString()
        }

        return null
    }
}